### PR TITLE
Support subresources

### LINKED ISSUE #22: Support subresources
Most resources are represented by a string representation of their name, such as "pods", just as it appears in the URL for the relevant API endpoint. However, some Kubernetes APIs involve a "subresource", such as the logs for a pod. The URL for the pods logs endpoint is:

```
GET /api/v1/namespaces/{namespace}/pods/{name}/log
```

In this case, “pods” is the namespaced resource, and “log” is a subresource of pods. To represent this in an RBAC role, use a slash to delimit the resource and subresource. To allow a subject to read both pods and pod logs, you would write:

```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: Role
metadata:
  namespace: default
  name: pod-and-pod-logs-reader
rules:
- apiGroups: [""]
  resources: ["pods", "pods/log"]
  verbs: ["get", "list"]
```

That said we should add the `subresource` flag to the who-can plugin:

```
# Check to see who can read pod logs
$ kubectl who-can get pods --subresource=log

# Check to see who can scale deployments
$ kubectl who-can update deployment --subresource=scale
```

### PR DESCRIPTION
**Pull request by danielpacak**
Resolves #22

### CODE REVIEW

File: pkg/cmd/list.go [hunk 1]
```diff
@@ -59,85 +67,107 @@ func NewCmdWhoCan() *cobra.Command {
...
 				return errors.New("please specify at least a verb and a resource type")
 			}
 			return nil
 		},
-		Run: func(cmd *cobra.Command, args []string) {
-			namespace, _, err := configFlags.ToRawKubeConfigLoader().Namespace()
-			if err != nil {
-				fmt.Printf("Error getting namespace from context config: %v\n", err)
-				os.Exit(1)
+		SilenceUsage: true,
```
Comment (lizrice) [inline pkg/cmd/list.go]: Is it deliberate to silence usage here?

File: pkg/cmd/list_test.go [hunk 2]
```diff
@@ -192,16 +342,16 @@ func TestWhoCan_policyRuleMatches(t *testing.T) {
 	}{
 		{
 			scenario: "A",
-			verb:     "get", resource: "service", resourceName: "",
+			verb:     "get", resource: "services", resourceName: "",
```
Comment (lizrice) [inline pkg/cmd/list_test.go]: Would it still work for the resource to be singular "service"?

File: pkg/cmd/list.go [hunk 3]
```diff
@@ -59,85 +67,107 @@ func NewCmdWhoCan() *cobra.Command {
   kubectl who-can create services -n foo
 
   # List who can get the service named "mongodb" in namespace "bar"
-  kubectl who-can get svc mongodb --namespace bar`,
+  kubectl who-can get svc mongodb --namespace bar
+
+  # List who can read pod logs
+  kubectl who-can get pods --subresource=log`,
```
Comment (lizrice) [inline pkg/cmd/list.go]: Is this consistent with how can-i does it? And do we also support something like "who-can get pods/log"?

File: pkg/cmd/resource_resolver.go [hunk 4]
```diff
@@ -25,29 +27,45 @@ func NewResourceResolver(client discovery.DiscoveryInterface) ResourceResolver {
 	}
 }
 
-func (rv *resourceResolver) Resolve(verbArg, resourceArg string) (v1.APIResource, error) {
-	resource, err := rv.resourceFor(resourceArg)
+func (rv *resourceResolver) Resolve(verbArg, resourceArg, subResource string) (v1.APIResource, error) {
+	resource, err := rv.resourceFor(resourceArg, subResource)
 	if err != nil {
-		return v1.APIResource{}, fmt.Errorf("resolving resource: %v", err)
+		return v1.APIResource{}, err
```
[the remark this answers is not in the document]
Comment (danielpacak) [reply pkg/cmd/resource_resolver.go]: It's been already wrapped in the whoCan.Check() method with the "resolving resource:" prefix. So the output looked a bit clumsy:

```
./kubectl-who-can get swimming-pools
Error: resolving resource: resolving resource: the server doesn't have a resource type "swimming-pools" 
```

File: pkg/cmd/list.go (hunk 1 above)
Comment (danielpacak) [reply pkg/cmd/list.go]: Yup. I wanted to be consistent with the built-in kubectl commands and existing plugins:

```
$ kubectl auth can-i go home
Warning: the server doesn't have a resource type 'home'
yes
$ kubectl get swimming-pools
error: the server doesn't have a resource type "swimming-pools"
$ ./kubectl-who-can go home
Error: resolving resource: the server doesn't have a resource type "home"
```

File: pkg/cmd/list_test.go (hunk 2 above)
Comment (danielpacak) [reply pkg/cmd/list_test.go]: Yes. `policyRuleMatcher` does the matching against already resolved `APIResource` which is always plural (e.g. `pods` or `pods/log`). Here I simply made test fixture reflect the implementation. The resolution itself is done by the `ResourceResolver` which supports plural, singular and short names. You can see that in the corresponding `TestResourceResolver_Resolve` test.

So to sum up the following works fine:

```
$ kubectl who-can get pods --subresource=log
$ kubectl who-can get pod --subresource=log
$ kubectl who-can get po --subresource=log
```

File: pkg/cmd/list.go (hunk 3 above)
Comment (danielpacak) [reply pkg/cmd/list.go]: Yes. The `can-i` subcommand has also the `--subresource` flag. We do not support `who-can get pods/logs`. You have to explicitly specify the subresource via the `--subresource` flag.

```
$ kubectl auth can-i get pods --subresource=log --as dpacak
yes
$ kubectl auth can-i get pods/logs --as dpacak
no
$ kubectl auth can-i get pods/log --as dpacak
no
$ kubectl who-can get pods --subresource=log
ROLEBINDING              NAMESPACE  SUBJECT  TYPE  SA-NAMESPACE
dpacak-can-get-pod-logs  default    dpacak   User

...
```

NB For `can-i` the invocation `kubect auth can-i get pod/logs` would mean check if I can get `pod` named `logs`. Whereas in `who-can` we have a third arg to specify the resource name. That is somehow inconsistent between the two.

Comment (lizrice) [reply pkg/cmd/list.go]: Interesting, I suppose we should get ourselves consistent with `can-i` for this `<resource type>/<resource name` invocation. But that can be a separate issue / PR.

**Pull request merged**